actors/miner/src/lib.rs

```diff
@@ -4904,13 +4916,97 @@ fn confirm_sector_proofs_valid_internal(
 
         Ok((total_pledge, newly_vested))
     })?;
-
     // Request pledge update for activated sector.
     notify_pledge_changed(rt, &(total_pledge - newly_vested))?;
 
     Ok(())
 }
 
+pub struct SectorActivation {
+    sector_number: u64,
+    seal_proof: RegisteredSealProof,
+    expiration: ChainEpoch,
+    sealed_cid: Cid,
+    data_activation: DataActivation,
+    pre_commit_deposit: Option<TokenAmount>,
+}
+
+impl SectorActivation {
+    // XXX: if we stop putting deal ids on sector infos this
+    // function can go away before merging back to master
+    fn deal_ids(&self) -> Vec<DealID> {
+        match &self.data_activation {
+            BuiltinActivation(deal_ids) => deal_ids.clone(),
+            //_ => vec![],
+        }
+    }
+}
+
+impl From<SectorPreCommitOnChainInfo> for SectorActivation {
+    fn from(si: SectorPreCommitOnChainInfo) -> SectorActivation {
+        SectorActivation {
+            sector_number: si.info.sector_number,
+            seal_proof: si.info.seal_proof,
+            expiration: si.info.expiration,
+            sealed_cid: si.info.sealed_cid,
+            data_activation: BuiltinActivation(si.info.deal_ids), // XXX distinguish data activation type from existence of deal_id / commD when generalizing
+            pre_commit_deposit: Some(si.pre_commit_deposit),
+        }
+    }
+}
+
+/*
+    BuiltinMarket is today's call in CSPV f05 -> f06, return deal weights as today
+    BuiltinMarketWithCommD is what PRU should do to avoid double the market calls, return deal weights as today
+    SectorContentAdded is the new thing and can include piece manifests, independent calls f05 (or other) and f06, return deal weights per spec
+    CC means its CC and so we don't need to do any data activation DealSpace is always zero
+*/
+enum DataActivation {
+    BuiltinActivation(Vec<DealID>),
+    //    BuiltinActivationCheckCommD(Vec<DealID>), // for use in PRU single shot activation and CommD check
+    //    SectorContentAdded(), // TODO add in allocids and piece manifest type to handle new activation flow
+    //    CC, // we can simplify activate and claim method by moving CC case directly here
+}
+
+fn activate_data(
+    rt: &impl Runtime,
+    activations: &[SectorActivation],
+) -> Result<(BatchReturn, Vec<DealSpaces>), ActorError> {
+    let (builtin_activation, builtin_activation_with_checks, sector_content_added, cc): (
+        Vec<DealsActivationInfo>,
+        Vec<()>, // XXX this is placeholder return types for batch data activation of different types
+        Vec<()>,
+        Vec<()>,
+    ) = activations.iter().fold((Vec::new(), Vec::new(), Vec::new(), Vec::new()), |mut acc, sa| {
```

Comment:
I don't love this. Definitely think we should constrain to a single activation method per call. We can probably use a single return type to represent any of them though?